### PR TITLE
Development

### PR DESCRIPTION
**Pull request by rvvergara**

### CODE REVIEW

File: app/controllers/v1/books_controller.rb [hunk 1]
```diff
@@ -0,0 +1,34 @@
+class V1::BooksController < ApplicationController
+  before_action :pundit_user
+  def index
+    @books = Book.all
+
+    render :books,status: :ok
```
Comment (nidalaa) [inline app/controllers/v1/books_controller.rb]: Space 👀

File: app/controllers/v1/collection_items_controller.rb [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+  before_action :pundit_user
+
+  def index
+    @user = User.find_by(username: params[:user_username])
+    @collection = @user.collection
+    render :index, status: :ok
+  end
+
+  def create
+    @item = @current_user.items.build(item_params)
```
Comment (nidalaa) [inline app/controllers/v1/collection_items_controller.rb]: No authorization needed?

File: app/controllers/v1/users_controller.rb [hunk 3]
```diff
@@ -1,11 +1,11 @@
 class V1::UsersController < ApplicationController
-  before_action :authenticate!, only: [:update, :show]
-  before_action only: [:update] do
-    allow_correct_user(User.find_by_username(params[:username])) unless @current_user.access_level > 2
+  before_action :pundit_user, only: [:update, :show, :index]
 
-    admin_allowed_change(User.find_by_username(params[:username])) unless @current_user.access_level < 3
+  def index
+    @users = User.all
+    render :users, locals: {users: @users},status: :ok
```
Comment (nidalaa) [inline app/controllers/v1/users_controller.rb]: Space 🔍

File: app/controllers/v1/users_controller.rb [hunk 4]
```diff
@@ -27,12 +27,14 @@ def create
   end
 
   def update
-    @user = User.find_by_username(params[:username])
-
-    if @user.update(user_params)
+    @user = User.find_by(id: params[:id])
+    authorize @user
+    if @user.update(permitted_attributes(@user)) && permitted_attributes(@user) != {}
```
Comment (nidalaa) [inline app/controllers/v1/users_controller.rb]: If you change order, you won't event start update if attributes hash is empty.
`permitted_attributes(@user) != {} && @user.update(permitted_attributes(@user))`

File: spec/models/book_spec.rb [hunk 5]
```diff
@@ -0,0 +1,37 @@
+require 'rails_helper'
+
+RSpec.describe Book, type: :model do
+  describe "validations" do
```
Comment (nidalaa) [inline spec/models/book_spec.rb]: Positive and negative scenario <3

File: spec/models/book_spec.rb [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+        harry.title = nil
+        harry.valid?
+        expect(harry.errors["title"]).to include("can't be blank")
+      end
+    end
+  end
+
+  describe "#search_by_term scope" do
+    before do
+      10.times do
```
Comment (nidalaa) [inline spec/models/book_spec.rb]: Do you really need so many objects to run this test?
The less object the better - not a problem now but when your app grows... Be scrappy fro the beginning :)

File: spec/models/collection_item_spec.rb [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+    context "no current page" do
+      it "is invalid" do
+        collection_item.current_page = nil
+        collection_item.valid?
+        expect(collection_item.errors["current_page"]).to include("can't be blank")
+      end
+    end
+
+    context "user wants to add same book to collection" do
+      it "is raises an error" do
```
Comment (nidalaa) [inline spec/models/collection_item_spec.rb]: it raises an error

File: spec/models/user_spec.rb [hunk 8]
```diff
@@ -35,4 +35,18 @@
       expect(arthur.username).to eq("arthur")
     end
   end
+
+  describe "#collection method" do
+    let(:john) { create(:user) }
+    before do
+      5.times do
```
Comment (nidalaa) [inline spec/models/user_spec.rb]: 2 would be enough, right?

File: spec/policies/book_policy_spec.rb [hunk 9]
```diff
@@ -0,0 +1,16 @@
+require 'rails_helper'
+
+RSpec.describe BookPolicy, type: :policy do
+  let(:admin) { create(:user, access_level: 2)}
+  let(:user) { create(:user, access_level: 1) }
+
+  context "admin book creation" do
```
Comment (nidalaa) [inline spec/policies/book_policy_spec.rb]: Maybe group it by action?

describe "book creation" do
  context "when logged-in as admin" do
  context  "when logged-in as regular user"

File: app/models/collection_item.rb [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+  belongs_to :book
+  belongs_to :user
+
+  validates :current_page, presence: true
+  validates :book_id, 
+            uniqueness: { scope: :user_id,
+                          message: "can only be added once in collection"
+                        } 
+
+  def book_data
```
Comment (nidalaa) [inline app/models/collection_item.rb]: Can you elaborate why do you need a collection item between users and books?

**Comment by nidalaa** [comment]
Hi @rvvergara,

I did a very quick and initial review :)
In general, it looks very promising and I really needed to look really carefully in order to find something to comment 🔍 

Please answer my question about the collection item model, as I am not sure what is its role.
Maybe there is something we can do better in that area.


To sum up:

awesome

File: app/models/collection_item.rb (hunk 10 above)
Comment (rvvergara) [reply app/models/collection_item.rb]: Thanks for this.

Ok The app is a BookStore CMS. We can think of the `Book` as the official supplier of a particular book and `CollectionItem` as the copy owned by a user. So the `admin` can buy more Books from the supplier and put in the app's library. Meanwhile, an ordinary user can search the library (Books already purchased from the supplier) and add a particular book in his/her collection (CollectionItem). 

The `Book` itself comes from a 3rd party API (which is called in the frontend - Googlebooks API).

I hope I was able to clarify it :)

File: spec/policies/book_policy_spec.rb (hunk 9 above)
Comment (rvvergara) [reply spec/policies/book_policy_spec.rb]: Got it! Thanks haha. It's my first time to write policy specs hahaha

File: spec/models/user_spec.rb (hunk 8 above)
Comment (rvvergara) [reply spec/models/user_spec.rb]: Haha yeah you are right. Ok i'll reduce the test data

File: spec/models/collection_item_spec.rb (hunk 7 above)
Comment (rvvergara) [reply spec/models/collection_item_spec.rb]: :+1:

File: spec/models/book_spec.rb (hunk 6 above)
Comment (rvvergara) [reply spec/models/book_spec.rb]: Haha ok yeah you are right. I'll tone it down to just 2 books

File: app/controllers/v1/users_controller.rb (hunk 4 above)
Comment (rvvergara) [reply app/controllers/v1/users_controller.rb]: Can you clarify what you mean by that? :)

File: app/controllers/v1/collection_items_controller.rb (hunk 2 above)
Comment (rvvergara) [reply app/controllers/v1/collection_items_controller.rb]: Nope... User only needs to be authenticated to add a "book" in his/her collection. Book objects are the ones that need admin access.

**Pull request merged**